### PR TITLE
Add OpenCL model

### PR DESCRIPTION
**Pull request by rory-improbable**
It's the big one, the one we've all been waiting for...

This PR adds all the code for the OpenCL model and UI, as well as a bit of refactoring so that the population initialisation and both models can be run from the same configurable `microsim/main.py` script.

### Current status
All the code is now in place, and both models seem to run fine from the new main script. The OpenCL model is producing different results from before, however I suspect that is because the population initialisation had already changed on the master branch with Nick's recent improvements (eg. the new workplaces), but it's worth investigating this properly. It seems like lockdown is no longer having an effect, I'm not sure if that is a bug I've introduced or if that is to be expected.

**update**
All the tests now pass, for both the PopulationInitialiser and Microsim class, as well as the OpenCL model tests. 

### New in this PR
- Added all OpenCL code in `microsim/opencl` folder.
- New `microsim/main.py` script to run both models, configurable with several different command line args (detailed in the readme) to choose which implementation to use, and whether to use the GUI for the OpenCL model or run in headless mode.
- Split the previous `Microsim` class into two new files
  - the population initialisation logic now lives in a new `PopulationInitialisation` class in `microsim/population_initialisation.py`. This returns the `individuals` and `activity_locations` dataframes so they can be passed to either model implementation.
  - the logic for timestepping the model still lives in the same `Microsim` class in `microsim/microsim_model.py`
- Caching: the population initialisation output can be cached to avoid re-running, which is controlled by passing the `--use-cache` arg to the main script.
- `SnapshotConverter` class - this takes the `individuals` and `activity_locations` dataframes and converts them into a `Snapshot` ready to be used for the OpenCL model. This conversion process involves things like combining the people flows for different activities into a single array, as well as assigning people's home locations based on OSM building data. This snapshot conversion is also read from a cached version, depending on the same `--use-cache` flag.
- Updated READMEs with new running instructions.
- OpenCL dashboard, located at `microsim/opencl/opencl_dashboard.py`. This is still a stripped down version of the original dashboard, however we can discuss what functionality is needed in this dashboard and I can implement that in a future PR. 

**NB** When splitting the existing Microsim class I had to move around a lot of different arguments to these classes, everything seems to run fine, but it's worth making sure that the Python/ R model runs as expected, and that the correct configuration is still being passed into the right places. I tried to change as little actual code as possible.

### CODE REVIEW

**Comment by nickmalleson** [comment]
@rory-improbable I can't believe you did this all in 1 afternoon! I'm really excited to start to experiment. By coincidence I run a 120-iteration 10-repetition model yesterday to start to have a look at the results of the change to workplaces, so can use that to check that the results are similar after this merge. 

Anyway, I think the tests are failing at the moment because of a problem with the imports. I added `scipy` and `pyopencl` to the environment and that fixed a couple of the errors, but I can't see easily how to include `imgui` - it may need installing with pip? Anyway I'm going to leave that one  to you or @Sparrow0hawk because I'll definitely break something :-)

**Review by nickmalleson — APPROVED** [review]

**Comment by rory-improbable** [comment]
It was more like a day and a bit, not an afternoon ;) but I'm just really keen to get everyone unblocked so we can take some results!

@nickmalleson @Sparrow0hawk there are installation instructions for both MacOS and Linux in the [OpenCL readme](https://github.com/Urban-Analytics/RAMP-UA/blob/add-opencl-model/microsim/opencl/README.md#installation), so we should be able to get this installed properly in the CI environment. 

For MacOS it should be as simple as installing all the pip requirements, however for linux I think we may have to run these lines to install one of the GUI dependencies called GLFW, however we don't need to run the GUI for the tests so maybe this isn't necessary:
```
sudo apt-get install libglfw3
sudo apt-get install libglfw3-dev
``` 
`imgui` should be able to be installed by pip I think, it should be in the requirements.txt, so hopefully that will be fixed with Alex's latest change.

File: microsim/opencl/requirements.txt
Comment (cshenton) [inline microsim/opencl/requirements.txt:1]: Is it intentional that we have separate requirements files here?

File: microsim/opencl/ramp/snapshot_convertor.py
Comment (cshenton) [reply microsim/opencl/ramp/snapshot_convertor.py:10]: Actually disregard this. We are consistent with it, it's just a slightly odd value (one 0 followed by 31 ones).

**Review by cshenton — CHANGES_REQUESTED** [review]
This all looks good. The only thing I'd like to see changed is to use `os.path.join` when you're constructing paths within the cache directory.

**Comment by rory-improbable** [comment]
@Sparrow0hawk I got the CI builds working, to do this on linux I followed your suggestion of installing opencl using a conda package called `pocl`, which includes binary dependencies.

I have also confirmed that the OpenCL model is definitely producing very different results from the Python/R model. However I don't think we should let that block merging this PR. The main thing is that the Python / R model results haven't changed. Once we can make sure that is the case we can merge this, then I can start investigating properly why the OpenCL results are different. This may involve implementing some of the additional functionality that we discussed at standup.

So the only things remaining to do before merging this PR I would say are the following:
- @nickmalleson checks that the Python / R model is still producing the same results as before, if not we can both investigate any differences
- I address Charlie's comments on the PR

**Comment by Sparrow0hawk** [comment]
@rory-improbable Great, i'm also skimming through the files in the PR but I think once we're happy with both your points above we can create a release and then merge and probably create v1.0.0 including opencl code.

**Comment by rory-improbable** [comment]
Sounds great, it would definitely be useful to get your feedback on the code.

Cheers,
Rory

File: README.md
Comment (rory-improbable) [reply README.md:72]: ahh yes I just tested and it turns out it does work, but **only** if you install the RAMP-UA package by running the setup script, as detailed above, eg.
```
python setup.py develop
```

File: README.md
Comment (rory-improbable) [reply README.md:77]: good call, made this change

File: microsim/opencl/ramp/snapshot.py
Comment (rory-improbable) [reply microsim/opencl/ramp/snapshot.py:142]: good call

File: microsim/opencl/ramp/kernels/ramp_ua.cl
Comment (rory-improbable) [reply microsim/opencl/ramp/kernels/ramp_ua.cl:225]: I'm aiming to avoid changing the OpenCL kernel code in this PR, to make it easier to assert that no functionality has changed, perhaps we can create a ticket to remind ourselves to do this in a subsequent PR?

File: microsim/opencl/requirements.txt
Comment (rory-improbable) [reply microsim/opencl/requirements.txt:1]: yes I think it's nice to keep this separate, as Alex has made the update which means the conda environment still installs this requirements.txt as well

File: microsim/initialisation_cache.py
Comment (rory-improbable) [reply microsim/initialisation_cache.py:10]: I'm actually not using this directory any more, I'll change this default to the cache directory that already existed and is gitignored

File: microsim/initialisation_cache.py
Comment (rory-improbable) [reply microsim/initialisation_cache.py:32]: so in theory this should never be reachable, because outside this I'm already checking if the file exists and if it doesn't then it just automatically re-runs the preprocessing. So I think I will just remove this check and if the code somehow does get here then it will throw an exception

File: model_parameters/default_dashboard.yml
Comment (Sparrow0hawk) [inline model_parameters/default_dashboard.yml:14]: Does changing this but not default.yml for the model lead to any problems when then loading the python dashboard? (I can't quite remember how these two files interact) but obviously want the scenario_dir folder to line up with scenario entry in `defaults.yml`

**Review by Sparrow0hawk — APPROVED** [review]
Bar addressing Charlie's comments I think this is good to go!

File: model_parameters/default_dashboard.yml
Comment (rory-improbable) [reply model_parameters/default_dashboard.yml:14]: ahh crap I didn't mean to change this file, the OpenCL dashboard doesn't even use this file, so it should be totally separate. I just changed this locally to show results from the python model. I'll revert this change!

File: README.md
Comment (cshenton) [reply README.md:72]: That's all good. So long as that onboarding flow is documented.

File: environment.yml
Comment (cshenton) [reply environment.yml:15]: Offline discussion, only mac/linux is officially supported so this is not an issue.

File: microsim/opencl/ramp/kernels/ramp_ua.cl
Comment (cshenton) [reply microsim/opencl/ramp/kernels/ramp_ua.cl:225]: Yep, good call.

**Review by cshenton — APPROVED** [review]
Thanks for making those changes. LGTM.

**Comment by nickmalleson** [comment]
@rory-improbable from what I can tell I think we're good to go. Did you want to make any other changes re. @cshenton 's comments? Otherwise I'll merge this and start to experiment :-) 

@spoonerf @jabrams23 are you OK if we update to the new version? The main thing you'll notice is that you need to call `main.py` rather than `microsim_model.py` but otherwise the arguments are basically the same. 

By the way I ran a scenario in python before the weekend, using a slightly earlier version of this branch, and the results were similar to those I ran in the pre-merge version.

File: microsim/main.py
Comment (nickmalleson) [inline microsim/main.py:57]: @rory-improbable I just noticed that some of the short versions of these parameters all share `-s`. Is this delibreate? I'm not sure what the behaviour will be as e.g. how do you distinguish between `@click.option('-s', '--use-cache/--dont-use-cache',` and `@click.option('-s', '--opencl/--no-opencl',` . If this is deliberate it's not a big deal, I can quickly change the duplictes but didn't want to in csae this is intentional.

**Comment by rory-improbable** [comment]
> @rory-improbable from what I can tell I think we're good to go. Did you want to make any other changes re. @cshenton 's comments? Otherwise I'll merge this and start to experiment :-)
> 
> @spoonerf @jabrams23 are you OK if we update to the new version? The main thing you'll notice is that you need to call `main.py` rather than `microsim_model.py` but otherwise the arguments are basically the same.
> 
> By the way I ran a scenario in python before the weekend, using a slightly earlier version of this branch, and the results were similar to those I ran in the pre-merge version.

@nickmalleson fantastic! Thanks for doing these checks.
I've addressed @cshenton's remaining comments and he has approved them this morning, so feel free to go ahead and merge!

File: microsim/main.py
Comment (rory-improbable) [reply microsim/main.py:57]: @nickmalleson ahh this is definitely not deliberate, I didn't notice this shorthand option!! If you could change the duplicates to something sensible that would be great

Comment (rory-improbable) [reply microsim/main.py:57]: good catch, this could definitely have caused some confusion

**Comment by Sparrow0hawk** [comment]
@nickmalleson are we ready to merge? If so i'll sort a pre-merge release and then you can merge this?

**Comment by nickmalleson** [comment]
> @nickmalleson are we ready to merge? If so i'll sort a pre-merge release and then you can merge this?

Yes please @Sparrow0hawk let me know when you're ready

**Comment by nickmalleson** [comment]
> > @nickmalleson are we ready to merge? If so i'll sort a pre-merge release and then you can merge this?
> 
> Yes please @Sparrow0hawk let me know when you're ready

Oh sorry @Sparrow0hawk actually give it 5 mins, I just want to check something minor with the parameters

File: microsim/main.py
Comment (nickmalleson) [reply microsim/main.py:57]: Sounds good, I'll do that now

**Comment by nickmalleson** [comment]
OK @Sparrow0hawk we're good to go :-) Let me know when I should merge

**Comment by spoonerf** [comment]
@nickmalleson All good with me, excited to play with it!

**Comment by jabrams23** [comment]
Awesome @nickmalleson <https://github.com/nickmalleson>.

**Comment by Sparrow0hawk** [comment]
@nickmalleson release is all done, if you could merge the release branch with master https://github.com/Urban-Analytics/RAMP-UA/pull/151 and then merge this branch we're all good.

Once this branch is merged i'll create a new release for v1.0.0

**Comment by nickmalleson** [comment]
> @nickmalleson release is all done, if you could merge the release branch with master #151 and then merge this branch we're all good.
> 
> Once this branch is merged i'll create a new release for v1.0.0

All done, version away!

**Pull request merged**
